src/value/mod.rs

```diff
@@ -0,0 +1,341 @@
+    type Struct: Struct + ?Sized;
+
+    /// The type of data this element represents.
+    fn ion_type(&self) -> IonType;
+
+    /// The annotations for this element.
+    ///
+    /// ## Usage
+    /// ```
+    /// # use ion_rs::value::*;
+    /// # use ion_rs::value::owned::*;
+    /// # use ion_rs::value::borrowed::*;
+    /// // simple function to extract the annotations to owned strings
+    /// fn annotation_strings<T: Element>(elem: &T) -> Vec<String> {
+    ///     elem.annotations().map(|tok| tok.text().unwrap().into()).collect()
+    /// }
+    ///
+    /// let strs = vec!["a", "b", "c"];
+    /// let owned_elem = OwnedElement::new(
+    ///     strs.iter().map(|s| (*s).into()).collect(),
+    ///     OwnedValue::String("moo".into())
+    /// );
+    /// let borrowed_elem = BorrowedElement::new(
+    ///     strs.iter().map(|s| (*s).into()).collect(),
+    ///     BorrowedValue::String("moo")
+    /// );
+    ///
+    /// let expected: Vec<String> = strs.iter().map(|s| (*s).into()).collect();
+    /// assert_eq!(expected, annotation_strings(&owned_elem));
+    /// assert_eq!(expected, annotation_strings(&borrowed_elem));
+    /// ```
+    ///
+    /// Note that this uses a `Box<dyn Iterator<...>>` to capture the borrow cleanly without
+    /// without [generic associated types (GAT)][gat].  In theory, when GAT lands, this could
+    /// be replaced with static polymorphism.
+    ///
+    /// [gat]: https://rust-lang.github.io/rfcs/1598-generic_associated_types.html
+    fn annotations<'a>(&'a self) -> Box<dyn Iterator<Item = &'a Self::SymbolToken> + 'a>;
+
+    /// Returns whether this element is a `null` value
+    fn is_null(&self) -> bool;
+
+    /// Returns a reference to the underlying [`AnyInt`] for this element.
+    ///
+    /// This will return `None` if the type is not `int` or the value is any `null`.
+    fn as_any_int(&self) -> Option<&AnyInt>;
+
+    /// Returns a slice to the textual value of this element.
+    ///
+    /// This will return `None` in the case that the type is not `string`/`symbol`,
+    /// if the value is any `null`, or the text of the `symbol` is not defined.
+    fn as_str(&self) -> Option<&str>;
+
+    /// Returns a reference to the [`SymbolToken`] of this element.
+    ///
+    /// This will return `None` in the case that the type is not `symbol` or the value is
+    /// any `null`.
+    fn as_sym(&self) -> Option<&Self::SymbolToken>;
+
+    /// Returns a reference to the [`Sequence`] of this element.
+    ///
+    /// This will return `None` in the case that the type is not `sexp`/`list` or
+    /// if the value is any `null`.
+    fn as_sequence(&self) -> Option<&Self::Sequence>;
+
+    /// Returns a reference to the [`Struct`] of this element.
+    ///
+    /// This will return `None` in the case that the type is not `struct` or the value is
+    /// any `null`.
+    fn as_struct(&self) -> Option<&Self::Struct>;
+
+    // TODO add all the accessors to the trait
```

Comment:
Thinking ahead for this: it would be nice to have `Element` also implement `IntAccess` so instead of:
```rust
let int: Option<i64> = element.as_any_int().as_i64();
```
callers could use:
```rust
let int: Option<i64> = element.as_i64();
```
as long as `element` is an `int`. (`Element` could still provide an `as_any_int` method, this would just be an ergonomic shortcut for a common case.)

Reply:
I missed that, thanks for pointing it out.